src/main/scala/metadata.scala

```diff
@@ -0,0 +1,83 @@
+package suffuse
+
+/** AKey is an attribute key: a Metadata map key which carries
+ *  both the type of the corresponding value, and a value to
+ *  provide when there is no value in the map. This as opposed
+ *  to wrapping everything in sight in Option.
+ */
+final class AKey[A](description: String, empty: => A) {
+  def emptyValue: A = empty
+  override def toString = description
+}
+
+/** Attr is a dependently typed key/value pair.
+ */
+trait Attr extends Any {
+  type Type
+  def key: AKey[Type]
+  def value: Type
+
+  def pair                               = key -> value
+  def matches[A](implicit that: AKey[A]) = key == that
+  def to_s                               = s"$key: $value"
+}
+
+/** The simple implementation of Attr, which ties the knot between the
+ *  user-facing type parameter and the type member which couples key and value.
+ */
+final case class AttrOf[A](key: AKey[A], value: A) extends Attr {
+  type Type = A
+  override def toString = s"$key: $value"
+}
+
+object Attr {
+  def apply[A](value: A)(implicit key: AKey[A]): AttrOf[A] = new AttrOf[A](key, value)
+  def unapply(x: Attr): Some[(AKey[x.Type], x.Type)]       = Some(x.pair)
+}
+
+/** A Metadata map, holding any number of well typed Attrs.
+ *  A typed value can be obtained for any key.
+ */
+final class Metadata(attrs: Vector[Attr]) {
+  private[this] val untypedMap                   = attrs.foldLeft(Map[AKey[_], Any]())(_ + _.pair)
+  private def doApply[A]()(implicit z: AKey[A]): A = untypedMap(z).asInstanceOf[A]
+
+  def isEmpty                                = untypedMap.isEmpty
+  def apply[A]()(implicit z: AKey[A]): A     = if (has[A]) doApply[A]() else z.emptyValue
+  def has[A]()(implicit z: AKey[A]): Boolean = keys contains z
+  def keys: Vector[AKey[_]]                  = attrs map (_.key)
+
+  /** Set could also hold onto the old value, and just return the last one.
+   *  It would amount to preserving the attribute's history.
+   */
+  def set(attr: Attr): Metadata        = new Metadata(attrs.filterNot(_.key eq attr.key) :+ attr)
+  def set[A: AKey](value: A): Metadata = set(Attr[A](value))
+
+  override def toString = if (isEmpty) "{ }" else attrs mkString ("{\n  ", "\n  ", "\n}")
+}
+object Metadata {
+  private[this] val Empty: Metadata = new Metadata(Vector())
+
+  def empty(): Metadata             = Empty
```

Comment:
To me, in Scala, `def` communicates: this might have a side effect. `val` reminds me that this should not have a side effect.

I don't understand the negative implication of "val means the empty instance can be recognized with a reference check".